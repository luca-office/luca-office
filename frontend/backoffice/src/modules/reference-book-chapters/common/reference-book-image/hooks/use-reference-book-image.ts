import * as React from "react"
import {useDispatch} from "react-redux"
import {NotificationSeverity} from "shared/enums"
import {useUpdateReferenceBookContent} from "shared/graphql/hooks"
import {AppNotification} from "shared/models"
import {Option, uploadBinary} from "shared/utils"
import {updateNotification, UpdateNotificationAction} from "../../../../../redux/actions/ui/common-ui-action"

export interface UseReferenceBookImageHook {
  readonly hideUpdateModal: () => void
  readonly isDeleteOrlyVisible: boolean
  readonly operationLoading: boolean
  readonly setIsDeleteOrlyVisible: (isVisible: boolean) => void
  readonly setViewerModalVisible: (isVisible: boolean) => void
  readonly showUpdateModal: () => void
  readonly updateImage: (file: File) => void
  readonly updateModalVisible: boolean
  readonly viewerModalVisible: boolean
}

export const useReferenceBookImage = (referenceBookChapterId: UUID, contentId: UUID): UseReferenceBookImageHook => {
  const dispatch = useDispatch()

  const [isDeleteOrlyVisible, setIsDeleteOrlyVisible] = React.useState<boolean>(false)
  const [updateModalVisible, setUpdateModalVisible] = React.useState<boolean>(false)
  const [viewerModalVisible, setViewerModalVisible] = React.useState<boolean>(false)
  const [uploadingImage, setUploadingImage] = React.useState<boolean>(false)

  const showUpdateModal = () => setUpdateModalVisible(true)
  const hideUpdateModal = () => setUpdateModalVisible(false)

  const {updateReferenceBookContent, updateReferenceBookContentLoading} = useUpdateReferenceBookContent(
    referenceBookChapterId
  )

  const updateImage = (file: File) => {
    setUploadingImage(true)
    uploadBinary(file)
      .then(binary =>
        updateReferenceBookContent(contentId, {
          imageBinaryFileId: binary.data.id
        })
      )
      .then(() => hideUpdateModal())
      .catch(() =>
        dispatch<UpdateNotificationAction>(
          updateNotification(
            Option.of<AppNotification>({
              severity: NotificationSeverity.Error,
              messageKey: "file__swap_file_type_image_failed"
            })
          )
        )
      )
      .finally(() => setUploadingImage(false))
  }

  return {
    hideUpdateModal,
    isDeleteOrlyVisible,
    operationLoading: updateReferenceBookContentLoading || uploadingImage,
    setIsDeleteOrlyVisible,
    showUpdateModal,
    updateImage,
    updateModalVisible,
    setViewerModalVisible,
    viewerModalVisible
  }
}
