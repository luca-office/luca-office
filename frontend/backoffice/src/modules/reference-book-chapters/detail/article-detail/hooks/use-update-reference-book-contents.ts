import {useApolloClient} from "@apollo/client"
import {
  ReferenceBookChapterQuery,
  ReferenceBookChapterQueryVariables
} from "shared/graphql/generated/ReferenceBookChapterQuery"
import {referenceBookChapterQuery} from "shared/graphql/queries"
import {ReferenceBookArticleContent} from "shared/models"
import {Option} from "shared/utils"

export interface UseUpdateReferenceBookContentsHook {
  readonly updateContentsInCache: (articleId: UUID, contentsList: Option<ReferenceBookArticleContent>[]) => void
}

export const useUpdateReferenceBookContents = (referenceBookChapterId: UUID): UseUpdateReferenceBookContentsHook => {
  const client = useApolloClient()

  const getReferenceBook = () => {
    try {
      return Promise.resolve(
        client.readQuery<ReferenceBookChapterQuery>({
          query: referenceBookChapterQuery,
          variables: {id: referenceBookChapterId}
        })
      )
    } catch (err) {
      return client
        .query<ReferenceBookChapterQuery, ReferenceBookChapterQueryVariables>({
          query: referenceBookChapterQuery,
          variables: {id: referenceBookChapterId},
          fetchPolicy: "network-only"
        })
        .then(result => result.data)
    }
  }

  const updateContentsInCache = (articleId: UUID, contentsList: Option<ReferenceBookArticleContent>[]) => {
    const contents = contentsList.reduce(
      (accumulator, contentOption) => contentOption.map(content => [...accumulator, content]).getOrElse(accumulator),
      [] as ReferenceBookArticleContent[]
    )
    getReferenceBook().then(
      data =>
        data?.referenceBookChapter &&
        client.writeQuery<ReferenceBookChapterQuery, ReferenceBookChapterQueryVariables>({
          query: referenceBookChapterQuery,
          variables: {id: referenceBookChapterId},
          data: {
            referenceBookChapter: {
              ...data!.referenceBookChapter,
              articles: data!.referenceBookChapter!.articles.map(article =>
                article.id === articleId ? {...article, contents} : article
              )
            }
          }
        })
    )
  }

  return {updateContentsInCache}
}
