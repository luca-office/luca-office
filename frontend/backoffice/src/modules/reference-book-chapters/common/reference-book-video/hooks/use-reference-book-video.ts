import * as React from "react"
import {useDispatch} from "react-redux"
import {NotificationSeverity} from "shared/enums"
import {useUpdateReferenceBookContent} from "shared/graphql/hooks"
import {AppNotification} from "shared/models"
import {Option, uploadBinary} from "shared/utils"
import {updateNotification, UpdateNotificationAction} from "../../../../../redux/actions/ui/common-ui-action"

export interface UseReferenceBookVideoHook {
  readonly hideUpdateModal: () => void
  readonly isDeleteOrlyVisible: boolean
  readonly operationLoading: boolean
  readonly setIsDeleteOrlyVisible: (isVisible: boolean) => void
  readonly setViewerModalVisible: (isVisible: boolean) => void
  readonly showUpdateModal: () => void
  readonly updateModalVisible: boolean
  readonly updateVideo: (file: File) => void
  readonly viewerModalVisible: boolean
}

export const useReferenceBookVideo = (referenceBookChapterId: UUID, contentId: UUID): UseReferenceBookVideoHook => {
  const dispatch = useDispatch()

  const [isDeleteOrlyVisible, setIsDeleteOrlyVisible] = React.useState<boolean>(false)
  const [updateModalVisible, setUpdateModalVisible] = React.useState<boolean>(false)
  const [viewerModalVisible, setViewerModalVisible] = React.useState<boolean>(false)
  const [uploadingVideo, setUploadingVideo] = React.useState<boolean>(false)

  const showUpdateModal = () => setUpdateModalVisible(true)
  const hideUpdateModal = () => setUpdateModalVisible(false)

  const {updateReferenceBookContent, updateReferenceBookContentLoading} = useUpdateReferenceBookContent(
    referenceBookChapterId
  )

  const updateVideo = (file: File) => {
    setUploadingVideo(true)
    uploadBinary(file)
      .then(binary =>
        updateReferenceBookContent(contentId, {
          videoBinaryFileId: binary.data.id
        })
      )
      .then(() => hideUpdateModal())
      .catch(() =>
        dispatch<UpdateNotificationAction>(
          updateNotification(
            Option.of<AppNotification>({
              severity: NotificationSeverity.Error,
              messageKey: "file__swap_file_type_video_failed"
            })
          )
        )
      )
      .finally(() => setUploadingVideo(false))
  }

  return {
    hideUpdateModal,
    isDeleteOrlyVisible,
    operationLoading: updateReferenceBookContentLoading || uploadingVideo,
    setIsDeleteOrlyVisible,
    setViewerModalVisible,
    showUpdateModal,
    updateModalVisible,
    updateVideo,
    viewerModalVisible
  }
}
