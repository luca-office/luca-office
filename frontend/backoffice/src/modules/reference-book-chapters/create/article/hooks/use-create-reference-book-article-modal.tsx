import {useForm, UseFormMethods} from "react-hook-form"
import {useDispatch} from "react-redux"
import {useCreateReferenceBookArticle} from "shared/graphql/hooks/mutations"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {Route} from "../../../../../routes"

export interface ReferenceBookArticleCreationForm {
  readonly title: string
}

export interface UseCreateReferenceBookHook {
  readonly createReferenceBookArticle: (creation: ReferenceBookArticleCreationForm) => void
  readonly dataLoading: boolean
  readonly formMethods: UseFormMethods<ReferenceBookArticleCreationForm>
  readonly dismissModal: () => void
}

export const useCreateReferenceBookArticleModal = (
  referenceBookChapterId: UUID,
  articleId?: UUID
): UseCreateReferenceBookHook => {
  const dispatch = useDispatch()
  const {createReferenceBookArticle, createReferenceBookArticleLoading: dataLoading} = useCreateReferenceBookArticle()
  const formMethods = useForm<ReferenceBookArticleCreationForm>()

  const dismissModal = () =>
    articleId
      ? dispatch(
          navigateToRouteAction(Route.ReferenceBookChaptersDetailArticle, {id: referenceBookChapterId, articleId})
        )
      : dispatch(navigateToRouteAction(Route.ReferenceBookChaptersDetail, {id: referenceBookChapterId}))

  const handleCreate = (creation: ReferenceBookArticleCreationForm) => {
    createReferenceBookArticle({...creation, referenceBookChapterId}, true).then(response =>
      response.forEach(data =>
        dispatch(
          navigateToRouteAction(Route.ReferenceBookChaptersDetailArticle, {
            id: data.referenceBookChapterId,
            articleId: data.id
          })
        )
      )
    )
  }

  return {
    createReferenceBookArticle: handleCreate,
    dataLoading,
    formMethods,
    dismissModal
  }
}
