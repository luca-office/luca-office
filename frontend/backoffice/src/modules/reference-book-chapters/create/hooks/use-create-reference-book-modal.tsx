import {useForm, UseFormMethods} from "react-hook-form"
import {useDispatch} from "react-redux"
import {ReferenceBookChapterCreation} from "shared/graphql/generated/globalTypes"
import {useCreateReferenceBookChapter} from "shared/graphql/hooks"
import {navigateToRouteAction} from "../../../../redux/actions/navigation-action"
import {Route} from "../../../../routes"

export interface ReferenceBookChapterCreationForm {
  readonly title: string
  readonly description: string
}

export interface UseCreateReferenceBookHook {
  readonly createReferenceBook: (creation: ReferenceBookChapterCreation) => void
  readonly createReferenceBookLoading: boolean
  readonly formMethods: UseFormMethods<ReferenceBookChapterCreationForm>
  readonly dismissModal: () => void
}

export const useCreateReferenceBookModal = (): UseCreateReferenceBookHook => {
  const dispatch = useDispatch()
  const {createReferenceBook, createReferenceBookLoading} = useCreateReferenceBookChapter()
  const formMethods = useForm<ReferenceBookChapterCreationForm>()

  const dismissModal = () => dispatch(navigateToRouteAction(Route.ReferenceBookChapters))

  const handleCreate = (creation: ReferenceBookChapterCreation) => {
    createReferenceBook(creation).then(response =>
      response.forEach(data => dispatch(navigateToRouteAction(Route.ReferenceBookChaptersDetail, {id: data.id})))
    )
  }

  return {
    createReferenceBook: handleCreate,
    createReferenceBookLoading,
    formMethods,
    dismissModal
  }
}
