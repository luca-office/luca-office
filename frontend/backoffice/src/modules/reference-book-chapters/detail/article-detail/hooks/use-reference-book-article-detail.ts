import {useMemo, useState} from "react"
import {useDispatch} from "react-redux"
import {TableContainerProps} from "shared/components"
import {ReferenceBookChapterUpdate, ReferenceBookContentType} from "shared/graphql/generated/globalTypes"
import {
  CreateReferenceBookContentProps,
  useCreateReferenceBookContent,
  useRepositionReferenceBookArticle,
  useRepositionReferenceBookContent,
  useUpdateReferenceBookArticle,
  useUpdateReferenceBookChapter,
  useUpdateReferenceBookContent
} from "shared/graphql/hooks/mutations"
import {ReferenceBookArticle, ReferenceBookArticleContent, ReferenceBookChapter} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {exists, find, Option, sort} from "shared/utils"
import {ResortableEntity, ResortedEntity} from "../../../../../models"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {Route} from "../../../../../routes"
import {getBookArticleTableProps} from "../../../config"
import {convertArticleToSortableArticleContents} from "../../../utils/reference-book-article-detail-utils"
import {useUpdateReferenceBookContents} from "./use-update-reference-book-contents"

export interface UseReferenceBookDetailHook
  extends Pick<CreateReferenceBookContentProps, "createReferenceBookContent"> {
  readonly articles: ReferenceBookArticle[]
  readonly articleTableProps: TableContainerProps<ReferenceBookArticle>
  readonly contentTypeModalVisible: boolean
  readonly dataLoading: boolean
  readonly hideSortModal: () => void
  readonly isPublished: boolean
  readonly navigateToChapterOverview: () => void
  readonly repositionArticleContent: (orderedArticleContent: ResortedEntity[]) => void
  readonly repositionReferenceBookContentLoading: boolean
  readonly selectContentType: (type: Option<ReferenceBookContentType>) => void
  readonly selectedArticle: Option<ReferenceBookArticle>
  readonly selectedContentType: Option<ReferenceBookContentType>
  readonly selectedEntityIsAnArticle: boolean
  readonly showSortModal: () => void
  readonly sortableArticleContents: Option<ResortableEntity[]>
  readonly sortModalVisible: boolean
  readonly title: string
  readonly toggleContentTypeModal: (visible: boolean) => void
  readonly toggleUpdateDescriptionModal: (visible: boolean) => void
  readonly updateArticleTitle: (title: string) => Promise<Option<ReferenceBookArticle>>
  readonly updateDescriptionModalVisible: boolean
  readonly updateReferenceBook: (update: Partial<ReferenceBookChapterUpdate>) => Promise<Option<ReferenceBookChapter>>
  readonly updateTextContent: (contentId: UUID, text: string) => Promise<Option<ReferenceBookArticleContent>>
}

export const useReferenceBookArticleDetail = (
  referenceBookChapter: ReferenceBookChapter,
  selectedEntityId: Option<UUID>,
  selectEntityId: (id: Option<UUID>) => void
): UseReferenceBookDetailHook => {
  const {t} = useLucaTranslation()
  const dispatch = useDispatch()

  const {updateReferenceBook, updateReferenceBookLoading} = useUpdateReferenceBookChapter()
  const {updateReferenceBookArticle, updateReferenceBookArticleLoading} = useUpdateReferenceBookArticle(
    referenceBookChapter.id
  )
  const {updateReferenceBookContent, updateReferenceBookContentLoading} = useUpdateReferenceBookContent(
    referenceBookChapter.id
  )
  const {createReferenceBookContent, createReferenceBookContentLoading} = useCreateReferenceBookContent(
    referenceBookChapter.id
  )
  const {repositionReferenceBookArticle, repositionReferenceBookArticleLoading} = useRepositionReferenceBookArticle(
    referenceBookChapter.id
  )
  const {repositionReferenceBookContent, repositionReferenceBookContentLoading} = useRepositionReferenceBookContent()
  const {updateContentsInCache} = useUpdateReferenceBookContents(referenceBookChapter.id)

  const [contentTypeModalVisible, toggleContentTypeModal] = useState(false)
  const [selectedContentType, selectContentType] = useState<Option<ReferenceBookContentType>>(Option.none())
  const [updateDescriptionModalVisible, toggleUpdateDescriptionModal] = useState(false)
  const [isArticleActionbarVisible, setIsArticleActionbarVisible] = useState<boolean>(false)
  const [sortModalVisible, setSortModalVisible] = useState<boolean>(false)

  const articles = useMemo<ReferenceBookArticle[]>(
    () => sort(article => article.position, referenceBookChapter.articles),
    [referenceBookChapter]
  )

  const selectedArticle = selectedEntityId.flatMap(id =>
    find(referenceBookArticle => referenceBookArticle.id === id, articles)
  )
  const title = selectedArticle.map(article => article.title).getOrElse(referenceBookChapter.title)

  const isPublished = !!referenceBookChapter.publishedAt

  const selectedEntityIsAnArticle = selectedEntityId
    .map(id => id !== referenceBookChapter.id && exists(article => article.id === id, articles))
    .getOrElse(false)

  const sortableArticleContents = convertArticleToSortableArticleContents(selectedArticle, t)

  const navigateToChapterOverview = () =>
    dispatch(navigateToRouteAction(Route.ReferenceBookChaptersDetail, {id: referenceBookChapter.id}))

  const navigateToArticleCreation = () =>
    dispatch(navigateToRouteAction(Route.ReferenceBookChaptersDetailArticleCreation, {id: referenceBookChapter.id}))

  const moveArticleUpwards = (article: ReferenceBookArticle) => {
    const currentIndex = articles.indexOf(article)
    repositionReferenceBookArticle(article.id, currentIndex === 1 ? undefined : articles[currentIndex - 2]?.id)
  }
  const moveArticleDownwards = (article: ReferenceBookArticle) => {
    const currentIndex = articles.indexOf(article)
    repositionReferenceBookArticle(article.id, articles[currentIndex + 1]?.id)
  }

  const handleUpdateBook = (update: Partial<ReferenceBookChapterUpdate>) =>
    updateReferenceBook(referenceBookChapter.id, {
      title: update.title || referenceBookChapter.title,
      description: update.description || referenceBookChapter.description
    })

  const handleUpdateArticleTitle = (title: string) =>
    selectedArticle.isDefined()
      ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        updateReferenceBookArticle(selectedArticle.map(article => article.id).orNull()!, {title})
      : Promise.resolve(Option.none<ReferenceBookArticle>())

  const handleUpdateTextContent = (contentId: UUID, text: string) => updateReferenceBookContent(contentId, {text})

  const showSortModal = () => setSortModalVisible(true)
  const hideSortModal = () => setSortModalVisible(false)

  const repositionArticleContent = (orderedArticleContent: ResortedEntity[]) => {
    Promise.all(orderedArticleContent.map(({id, predecessorId}) => repositionReferenceBookContent(id, predecessorId)))
      .then(contents => selectedArticle.forEach(article => updateContentsInCache(article.id, contents)))
      .then(hideSortModal)
  }

  const articleTableProps = getBookArticleTableProps({
    articles,
    isArticleActionbarVisible,
    isPublished,
    moveArticleDownwards,
    moveArticleUpwards,
    navigateToArticleCreation,
    referenceBookChapterId: referenceBookChapter.id,
    repositionArticleLoading: repositionReferenceBookArticleLoading,
    selectEntityId,
    setIsArticleActionbarVisible,
    t
  })

  return {
    articles,
    articleTableProps,
    contentTypeModalVisible,
    createReferenceBookContent,
    dataLoading:
      createReferenceBookContentLoading ||
      repositionReferenceBookArticleLoading ||
      repositionReferenceBookContentLoading ||
      updateReferenceBookArticleLoading ||
      updateReferenceBookContentLoading ||
      updateReferenceBookLoading,
    hideSortModal,
    isPublished,
    navigateToChapterOverview,
    repositionArticleContent,
    repositionReferenceBookContentLoading,
    selectContentType,
    selectedArticle,
    selectedContentType,
    selectedEntityIsAnArticle,
    showSortModal,
    sortableArticleContents,
    sortModalVisible,
    title,
    toggleContentTypeModal,
    toggleUpdateDescriptionModal,
    updateArticleTitle: handleUpdateArticleTitle,
    updateDescriptionModalVisible,
    updateReferenceBook: handleUpdateBook,
    updateTextContent: handleUpdateTextContent
  }
}
