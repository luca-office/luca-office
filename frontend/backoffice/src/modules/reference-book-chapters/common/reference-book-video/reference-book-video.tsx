import {css} from "@emotion/react"
import partial from "lodash-es/partial"
import * as React from "react"
import {DeleteOrArchiveEntityButton, Icon, Overlay, VideoViewer} from "shared/components"
import {BinaryType, IconName} from "shared/enums"
import {useDeleteReferenceBookContent} from "shared/graphql/hooks"
import {BinaryFile} from "shared/models"
import {backgroundColorDarker, Flex, spacing, spacingSmall} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {Option} from "shared/utils"
import {BinaryUpdateModal} from "../../../../components"
import {useReferenceBookVideo} from "./hooks/use-reference-book-video"

export interface ReferenceBookVideoProps {
  readonly contentId: UUID
  readonly readonly?: boolean
  readonly referenceBookChapterId: UUID
  readonly scrollId?: string
  readonly videoUrl: string
  readonly videoFile: Option<BinaryFile>
}

export const ReferenceBookVideo: React.FC<ReferenceBookVideoProps> = ({
  contentId,
  readonly,
  referenceBookChapterId,
  scrollId,
  videoUrl,
  videoFile
}) => {
  const {t} = useLucaTranslation()
  const {
    hideUpdateModal,
    operationLoading,
    showUpdateModal,
    updateVideo,
    updateModalVisible,
    isDeleteOrlyVisible,
    setIsDeleteOrlyVisible,
    setViewerModalVisible,
    viewerModalVisible
  } = useReferenceBookVideo(referenceBookChapterId, contentId)

  return (
    <React.Fragment>
      <div css={styles.container(isDeleteOrlyVisible)} data-scroll={scrollId}>
        <div css={styles.videoWrapper}>
          <video role="video" css={styles.video} src={videoUrl} controls={false} />
        </div>
        {!readonly ? (
          <div className={"edit-overlay"} css={styles.editOverlay} onClick={showUpdateModal}>
            <Icon customStyles={styles.control} color={"white"} name={IconName.EditBordered} />
            <div>{t("edit__video")}</div>
            <DeleteOrArchiveEntityButton
              customStyles={styles.control}
              stopEventPropagation={true}
              entityId={contentId}
              useDeleteHook={partial(useDeleteReferenceBookContent, referenceBookChapterId)}
              modalTitleKey={"reference_books__delete_dialog_title_video"}
              modalTextKey={"reference_books__delete_dialog_text_video"}
              isConfirmDialogVisible={isDeleteOrlyVisible}
              toggleIsConfirmDialogVisible={setIsDeleteOrlyVisible}
              stopEventPropagationOnOverlayClick={true}
            />
          </div>
        ) : (
          <div
            className={"edit-overlay"}
            css={[styles.editOverlay, styles.previewOverlay]}
            onClick={() => setViewerModalVisible(true)}>
            <div>{t("preview__video")}</div>
          </div>
        )}
        {updateModalVisible && (
          <BinaryUpdateModal
            type={BinaryType.Video}
            src={videoUrl}
            onConfirm={updateVideo}
            onDismiss={hideUpdateModal}
            disabled={operationLoading}
          />
        )}
      </div>
      {viewerModalVisible &&
        videoFile
          .map(video => (
            <Overlay>
              <VideoViewer
                binaries={[{id: video.id, title: video.filename, path: videoUrl}]}
                onClose={() => setViewerModalVisible(false)}
                customStyles={styles.viewer}
              />
            </Overlay>
          ))
          .orNull()}
    </React.Fragment>
  )
}

const Size = {
  video: 144
}

const styles = {
  container: (isDeleteOrlyVisible: boolean) =>
    css({
      position: "relative",
      display: "inline-block",
      width: Size.video,
      height: Size.video,
      margin: spacing(0, spacingSmall, spacingSmall, 0),
      backgroundColor: backgroundColorDarker,
      overflow: "hidden",

      ...(!isDeleteOrlyVisible && {
        "&:not(:hover) > .edit-overlay": {
          display: "none"
        }
      })
    }),
  videoWrapper: css(Flex.row, {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: "center",
    width: "100%",
    height: "100%"
  }),
  video: css({
    maxWidth: "100%",
    maxHeight: "100%"
  }),
  editOverlay: css(Flex.column, {
    position: "absolute",
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    backgroundColor: "rgba(46, 48, 50, 0.48)",
    color: "white",
    justifyContent: "space-between",
    alignItems: "center",
    cursor: "pointer"
  }),
  previewOverlay: css({
    justifyContent: "center"
  }),
  viewer: css({
    width: "75vw",
    height: "90vh"
  }),
  control: css({
    alignSelf: "flex-end",
    margin: spacingSmall
  })
}
