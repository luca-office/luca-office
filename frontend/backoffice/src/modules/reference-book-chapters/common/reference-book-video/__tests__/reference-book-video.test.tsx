import {MockedProvider} from "@apollo/client/testing"
import {act, render, screen} from "@testing-library/react"
import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {referenceBookChapterMock} from "shared/__mocks__"
import {VideoViewer} from "shared/components"
import {videoBinariesMock} from "shared/components/binary-viewer/__mocks__/binaries.mock"
import {MimeType} from "shared/graphql/generated/globalTypes"
import {deleteReferenceBookContentMutation} from "shared/graphql/mutations"
import {BinaryFile} from "shared/models"
import {Option} from "shared/utils"
import wait from "waait"
import {BinaryUpdateModal} from "../../../../../components"
import * as useReferenceBookVideoHook from "../hooks/use-reference-book-video"
import {UseReferenceBookVideoHook} from "../hooks/use-reference-book-video"
import {ReferenceBookVideo, ReferenceBookVideoProps} from "../reference-book-video"
import "@testing-library/jest-dom"

const referenceBookChapter = referenceBookChapterMock()
const referenceBookArticle = referenceBookChapter.articles[0]
const referenceBookContent = {
  ...referenceBookArticle.contents[0],
  imageBinaryFileId: null,
  videoBinaryFileId: "a39fcd97-942b-4750-a5a8-05e82565b0f2",
  videoUrl: "https://luca-develop.s3.eu-central-1.amazonaws.com/3dd596c7-b766-4fb2-9a8d-7b4a0a318c0e",
  imageBinaryFile: null,
  videoBinaryFile: {
    __typename: "BinaryFile",
    id: "a39fcd97-942b-4750-a5a8-05e82565b0f2",
    createdAt: new Date(2020, 10, 5).toISOString(),
    modifiedAt: new Date(2020, 10, 15).toISOString(),
    filename: "mov_bbb",
    fileSize: 1,
    mimeType: MimeType.VideoMp4,
    url: "https://luca-develop.s3.eu-central-1.amazonaws.com/3dd596c7-b766-4fb2-9a8d-7b4a0a318c0e"
  }
}

const defaultProps: ReferenceBookVideoProps = {
  contentId: referenceBookContent.id,
  videoUrl: referenceBookContent.videoUrl ?? "",
  readonly: false,
  referenceBookChapterId: referenceBookChapter.id,
  videoFile: Option.of(videoBinariesMock[0] as unknown as BinaryFile)
}

const hookValuesDefault: UseReferenceBookVideoHook = {
  hideUpdateModal: jest.fn(),
  operationLoading: false,
  showUpdateModal: jest.fn(),
  updateVideo: jest.fn(),
  updateModalVisible: false,
  isDeleteOrlyVisible: false,
  setIsDeleteOrlyVisible: jest.fn(),
  viewerModalVisible: false,
  setViewerModalVisible: jest.fn
}

const stateSpy = jest.spyOn(useReferenceBookVideoHook, "useReferenceBookVideo")

const getComponent = (props?: Partial<ReferenceBookVideoProps>) => (
  <MockedProvider
    mocks={[
      {
        request: {
          query: deleteReferenceBookContentMutation,
          variables: {
            id: referenceBookContent.id
          }
        },
        result: {
          data: {
            deleteReferenceBookContent: referenceBookContent.id
          }
        }
      }
    ]}
    addTypename={true}>
    <ReferenceBookVideo {...{...defaultProps, ...props}} />
  </MockedProvider>
)

describe("reference-book-video", () => {
  it("renders correctly", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = create(getComponent())
    await act(() => wait(0))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const tree = shallow(getComponent())
    render(getComponent())
    await act(() => wait(0))

    const referenceBookVideo = tree.find(ReferenceBookVideo)
    const content = referenceBookVideo.dive()

    const video = screen.getByRole("video") as HTMLVideoElement

    expect(video).toBeInTheDocument()
    expect(video.src).toContain(defaultProps.videoUrl)

    expect(content.find(".edit-overlay")).toHaveLength(1)
    expect(content.find(BinaryUpdateModal)).toHaveLength(0)
  })
  it("has correct structure (readonly)", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const tree = shallow(getComponent({readonly: true}))

    await act(() => wait(0))

    const referenceBookVideo = tree.find(ReferenceBookVideo)
    const content = referenceBookVideo.dive()

    expect(content.find(".edit-overlay")).toHaveLength(1)
    expect(content.find(BinaryUpdateModal)).toHaveLength(0)
  })
  it("has correct structure (update-modal visible)", async () => {
    stateSpy.mockReturnValue({...hookValuesDefault, updateModalVisible: true})
    const tree = shallow(getComponent())

    await act(() => wait(0))

    const referenceBookVideo = tree.find(ReferenceBookVideo)
    const content = referenceBookVideo.dive()

    expect(content.find(".edit-overlay")).toHaveLength(1)
    expect(content.find(BinaryUpdateModal)).toHaveLength(1)
  })
  it("has correct structure (viewer-modal visible)", async () => {
    stateSpy.mockReturnValue({...hookValuesDefault, viewerModalVisible: true})
    const tree = shallow(getComponent({readonly: true}))

    await act(() => wait(0))

    const referenceBookVideo = tree.find(ReferenceBookVideo)
    const content = referenceBookVideo.dive()

    expect(content.find(".edit-overlay")).toHaveLength(1)
    expect(content.find(VideoViewer)).toHaveLength(1)
  })
})
