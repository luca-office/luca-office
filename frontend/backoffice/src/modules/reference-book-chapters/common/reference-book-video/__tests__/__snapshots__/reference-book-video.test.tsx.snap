// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reference-book-video renders correctly 1`] = `
<div
  className="css-1qtaw86-ReferenceBookVideo"
>
  <div
    className="css-140lpvm-ReferenceBookVideo"
  >
    <video
      className="css-1ftwx9h-ReferenceBookVideo"
      controls={false}
      role="video"
      src="https://luca-develop.s3.eu-central-1.amazonaws.com/3dd596c7-b766-4fb2-9a8d-7b4a0a318c0e"
    />
  </div>
  <div
    className="edit-overlay css-ff7td4-ReferenceBookVideo"
    onClick={[MockFunction]}
  >
    <div
      className="css-1wr1ssb-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
    />
    <div>
      edit__video
    </div>
    <div
      className="css-1806nni-OrlyButton"
    >
      <button
        className="css-hnekdl-Button"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="css-95j2hm-Button"
        >
          <div
            className="css-mj8vo1-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
      </button>
    </div>
  </div>
</div>
`;
